File: DevExpress-Demo/DevExpress.ProductsDemo.Wpf/CS/DevExpress.ProductsDemo.Wpf/ModulesViewModels/SalesDemo/DateTimeUtils.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;

namespace ProductsDemo {
    public static class DateTimeUtils {
        public static DateTimeRange GetDayRange(DateTime date) {
            DateTime startOfDate = new DateTime(date.Year, date.Month, date.Day);
            DateTime endOfToday = startOfDate.AddDays(1).AddTicks(-1);
            return new DateTimeRange(startOfDate, endOfToday);
        }
        public static DateTimeRange GetLastWeekRange() {
            DayOfWeek firstDay = CultureInfo.CurrentCulture.DateTimeFormat.FirstDayOfWeek;
            DateTime today = DateTime.Today;
            DateTime startOfWeek = today;
            while(startOfWeek.DayOfWeek != firstDay) {
                startOfWeek = startOfWeek.AddDays(-1);
            }
            DateTime endOfWeek = startOfWeek.AddDays(7).AddTicks(-1);
            return new DateTimeRange(startOfWeek, endOfWeek);
        }
        public static DateTimeRange GetMonthRange(DateTime date) {
            DateTime startOfMonth = new DateTime(date.Year, date.Month, 1);
            int daysInCurrentMonth = DateTime.DaysInMonth(date.Year, date.Month);
            DateTime endOfMonth = startOfMonth.AddDays(daysInCurrentMonth).AddTicks(-1);
            return new DateTimeRange(startOfMonth, endOfMonth);
        }
        public static DateTimeRange GetYtdRange() {
            DateTime today = DateTime.Today;
            DateTime startOfYear = new DateTime(today.Year, 1, 1);
            DateTime endOfYear = new DateTime(today.Year, 12, 31);
            return new DateTimeRange(startOfYear, endOfYear);
        }
        public static DateTimeRange GetOneYearRange() {
            return new DateTimeRange(DateTime.Today.AddYears(-1), DateTime.Today);
        }

        public static DateTimeRange GetLastYearRange() {
            DateTime dateInLastYear = DateTime.Today.AddYears(-1);
            DateTime startOfYear = new DateTime(dateInLastYear.Year, 1, 1);
            DateTime endOfYear = startOfYear.AddYears(1).AddTicks(-1);
            return new DateTimeRange(startOfYear, endOfYear);
        }
        public static int GetLastYear() {
            return DateTime.Today.AddYears(-1).Year;
        }
        public static object GetCurrentYear() {
            return DateTime.Now.Year;
        }
    }

    public struct DateTimeRange {
        DateTime start;
        DateTime end;

        public DateTime Start {
            get { return start; }
        }
        public DateTime End {
            get { return end; }
        }

        public DateTimeRange(DateTime start, DateTime end) {
            this.start = start;
            this.end = end;
        }
    }

    public struct DecimalRange {
        decimal start;
        decimal end;

        public decimal Start {
            get { return start; }
        }
        public decimal End {
            get { return end; }
        }

        public DecimalRange(decimal start, decimal end) {
            this.start = start;
            this.end = end;
        }
    }

    public static class SalesRangeProvider {
        public static DecimalRange GetBadSalesRange() {
            return new DecimalRange(0, 200000000M);
        }
        public static DecimalRange GetNormalSalesRange() {
            return new DecimalRange(0, 400000000M);
        }
        public static DecimalRange GetGoodSalesRange() {
            return new DecimalRange(0, 600000000M);
        }
    }
}

File: DevExpress-Demo/DevExpress.ProductsDemo.Wpf/VB/DevExpress.ProductsDemo.Wpf/ModulesViewModels/SalesDemo/DateTimeUtils.vb
Imports System
Imports System.Globalization

Namespace ProductsDemo

    Public Module DateTimeUtils

        Public Function GetDayRange(ByVal [date] As Date) As DateTimeRange
            Dim startOfDate As Date = New DateTime([date].Year, [date].Month, [date].Day)
            Dim endOfToday As Date = startOfDate.AddDays(1).AddTicks(-1)
            Return New DateTimeRange(startOfDate, endOfToday)
        End Function

        Public Function GetLastWeekRange() As DateTimeRange
            Dim firstDay As DayOfWeek = CultureInfo.CurrentCulture.DateTimeFormat.FirstDayOfWeek
            Dim today As Date = Date.Today
            Dim startOfWeek As Date = today
            While startOfWeek.DayOfWeek <> firstDay
                startOfWeek = startOfWeek.AddDays(-1)
            End While

            Dim endOfWeek As Date = startOfWeek.AddDays(7).AddTicks(-1)
            Return New DateTimeRange(startOfWeek, endOfWeek)
        End Function

        Public Function GetMonthRange(ByVal [date] As Date) As DateTimeRange
            Dim startOfMonth As Date = New DateTime([date].Year, [date].Month, 1)
            Dim daysInCurrentMonth As Integer = Date.DaysInMonth([date].Year, [date].Month)
            Dim endOfMonth As Date = startOfMonth.AddDays(daysInCurrentMonth).AddTicks(-1)
            Return New DateTimeRange(startOfMonth, endOfMonth)
        End Function

        Public Function GetYtdRange() As DateTimeRange
            Dim today As Date = Date.Today
            Dim startOfYear As Date = New DateTime(today.Year, 1, 1)
            Dim endOfYear As Date = New DateTime(today.Year, 12, 31)
            Return New DateTimeRange(startOfYear, endOfYear)
        End Function

        Public Function GetOneYearRange() As DateTimeRange
            Return New DateTimeRange(Date.Today.AddYears(-1), Date.Today)
        End Function

        Public Function GetLastYearRange() As DateTimeRange
            Dim dateInLastYear As Date = Date.Today.AddYears(-1)
            Dim startOfYear As Date = New DateTime(dateInLastYear.Year, 1, 1)
            Dim endOfYear As Date = startOfYear.AddYears(1).AddTicks(-1)
            Return New DateTimeRange(startOfYear, endOfYear)
        End Function

        Public Function GetLastYear() As Integer
            Return Date.Today.AddYears(-1).Year
        End Function

        Public Function GetCurrentYear() As Object
            Return Date.Now.Year
        End Function
    End Module

    Public Structure DateTimeRange

        Private startField As Date

        Private endField As Date

        Public ReadOnly Property Start As Date
            Get
                Return startField
            End Get
        End Property

        Public ReadOnly Property [End] As Date
            Get
                Return endField
            End Get
        End Property

        Public Sub New(ByVal start As Date, ByVal [end] As Date)
            startField = start
            endField = [end]
        End Sub
    End Structure

    Public Structure DecimalRange

        Private startField As Decimal

        Private endField As Decimal

        Public ReadOnly Property Start As Decimal
            Get
                Return startField
            End Get
        End Property

        Public ReadOnly Property [End] As Decimal
            Get
                Return endField
            End Get
        End Property

        Public Sub New(ByVal start As Decimal, ByVal [end] As Decimal)
            startField = start
            endField = [end]
        End Sub
    End Structure

    Public Module SalesRangeProvider

        Public Function GetBadSalesRange() As DecimalRange
            Return New DecimalRange(0, 200000000D)
        End Function

        Public Function GetNormalSalesRange() As DecimalRange
            Return New DecimalRange(0, 400000000D)
        End Function

        Public Function GetGoodSalesRange() As DecimalRange
            Return New DecimalRange(0, 600000000D)
        End Function
    End Module
End Namespace
